### PR TITLE
new module: quotes

### PR DESCRIPTION
**Pull request by pqhaz**

### CODE REVIEW

File: quotes.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
+# -*- coding: future_fstrings -*-
+
+#    Quotes module for Friendly-Telegram
+#    Copyright (C) 2019 Pavel (pqhaz)
```
Comment (penn5) [inline quotes.py]: Just put the authors

File: quotes.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
+# -*- coding: future_fstrings -*-
+
+#    Quotes module for Friendly-Telegram
```
Comment (penn5) [inline quotes.py]: Don't customise the copyright header

File: quotes.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import logging
+import requests
+import ujson
```
Comment (penn5) [inline quotes.py]: No

File: quotes.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+def register(cb):
+    cb(QuotesMod())
+
+
+class QuotesMod(loader.Module):
+    """Quote a message."""
+    def __init__(self):
+        self.config = loader.ModuleConfig("API_TOKEN", "", "API Key/Token for Quotes.",
+                                          "API_URL", "http://api.antiddos.systems", "API URL for Quotes.")
+        self.name = _("Quotes")
```
Comment (penn5) [inline quotes.py]: This is deprecated. Use the new format for strings. This applies to all strings in the file.

File: quotes.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            msg = await message.edit('<code>Processing...</code>')
+
+        username = ''
+        no_pfp = True
+        colour = '#b48bf2'
+        admintitle = ''
+        username = ''
+        pfp = ''
+        user_id = 000000000
+        fwd_status = 'no_fwd'
```
Comment (penn5) [inline quotes.py]: Don't use this variable. Just do it inline.

File: quotes.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            admintitle = ''
+        else:
+            try:
+                admintitle = participant.participant.rank
+                if not admintitle:
+                    admintitle = 'admin'
+            except (AttributeError, UnboundLocalError):
+                admintitle = ''
+
+        data = ujson.dumps({
```
Comment (penn5) [inline quotes.py]: Use stdlib

File: quotes.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            "ProfilePhotoURL": pfp,
+            "UsernameColor": colour,
+            "Username": username,
+            "AdminTitle": admintitle,
+            "Text": reply.raw_text,
+            "Markdown": get_markdown(reply),
+            "Template": args[0].lower(),
+            "APIKey": self.config["API_TOKEN"]
+        })
+        requested = requests.post(f'{self.config["API_URL"]}/api/v2/quote', data=data)
```
Comment (penn5) [inline quotes.py]: NO NO NO NO NO, ASYNC

File: quotes.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    if not reply.entities:
+        return []
+
+    markdown = []
+    for entity in reply.entities:
+        entity_type = type(entity)
+        md_item = {
+            "Type": "",
+            "Start": entity.offset,
+            "End": entity.offset + (entity.length - 1)
```
Comment (penn5) [inline quotes.py]: Brackets are redundant

File: quotes.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+        }
+        if entity_type is MessageEntityBold:
+            md_item["Type"] = "bold"
+        elif entity_type is MessageEntityItalic:
+            md_item["Type"] = "italic"
+        elif entity_type in [MessageEntityMention, MessageEntityTextUrl,
```
Comment (penn5) [inline quotes.py]: Use () not []

File: quotes.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        file.close()
+
+
+def get_markdown(reply):
+    if not reply.entities:
+        return []
+
+    markdown = []
+    for entity in reply.entities:
+        entity_type = type(entity)
```
Comment (penn5) [inline quotes.py]: No. Use isinstance (built-in function)

File: quotes.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+import logging
+import requests
+import ujson
+import base64
+
+from .. import loader, utils
+from PIL import Image
+from io import BytesIO
+from urllib.error import HTTPError
+from telethon.errors.rpcerrorlist import ChannelPrivateError
```
Comment (penn5) [inline quotes.py]: Do full-module imports

File: quotes.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+        self.config = loader.ModuleConfig("API_TOKEN", "", "API Key/Token for Quotes.",
+                                          "API_URL", "http://api.antiddos.systems", "API URL for Quotes.")
+        self.name = _("Quotes")
+
+    async def client_ready(self, client, db):
+        self.client = client
+
+    async def quotecmd(self, message):  # noqa: C901
+        """Quote a message.
+        Usage: .quote (template) (optional: file/force_file/sticker)
```
Comment (penn5) [inline quotes.py]: Usage: .quote [template] [format]
Format can be file, force_file or sticker, defaulting to sticker.
If template is missing, possible templates are fetched.

File: quotes.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        Sends as sticker by default
+        Needs an API Token."""
+
+        args = utils.get_args(message)
+        reply = await message.get_reply_message()
+
+        if not reply:
+            return await utils.answer(message, '<code>You didn\'t reply to a message.</code>')
```
Comment (penn5) [inline quotes.py]: We use " not '. This applies globally.

File: quotes.py [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+        if not args:
+            return await utils.answer(message, '<code>You didn\'t specify template</code>')
+
+        if reply.media:
+            msg = await message.edit('<code>Processing... Note: only message will get processed.</code>')
```
Comment (penn5) [inline quotes.py]: Don't put processing messages. Also don't use message.edit

File: quotes.py [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+            await self.client.download_profile_photo(user_id, file=pfp_bytes)
+            pfp_bytes.seek(0)
+            if no_pfp is False:
+                pfp = 'data:image/png;base64, ' + base64.b64encode(pfp_bytes.read()).decode()
+            else:
+                pfp = ''
+            pfp_bytes.close()
+
+        if fwd_status not in ['hidden', 'channel_private']:
+            colors = ["#fb6169", "#85de85", "#f3bc5c", "#65bdf3", "#b48bf2", "#ff5694", "#62d4e3", "#faa357"]
```
Comment (penn5) [inline quotes.py]: Put these in order of the check digit so that double indexing isn't needed

Comment (penn5) [reply quotes.py]: Also make them configuration items

File: quotes.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+            num1 = user_id % 7
+            num2 = [0, 7, 4, 1, 6, 3, 5]
+            colour = colors[num2[num1]]
+        else:
+            colour = '#b48bf2'
```
Comment (penn5) [inline quotes.py]: Make it a config item

File: quotes.py [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+        if fwd_status == 'channel_open':
+            username = chat.title
+        elif fwd_status == 'hidden':
+            username = reply.fwd_from.from_name
+        elif fwd_status == 'channel_private':
+            username = 'Hidden.'
+        else:
+            username = user.first_name
+            if user.last_name:
```
Comment (penn5) [inline quotes.py]: Use telethon.utils.get_display_name

File: quotes.py [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+                username += f' {user.last_name}'
+
+        if fwd_status.startswith('channel'):
```
Comment (penn5) [inline quotes.py]: This code smells.

File: quotes.py [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+            admintitle = 'channel'
+        elif fwd_status == 'hidden':
+            admintitle = ''
+        else:
+            try:
+                admintitle = participant.participant.rank
+                if not admintitle:
+                    admintitle = 'admin'
+            except (AttributeError, UnboundLocalError):
```
Comment (penn5) [inline quotes.py]: Woah woah hold your horses. Wth is going on here? You should use getattr(participant, "participant", None) if you are worried there's no participant

File: quotes.py [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+                admintitle = ''
+
+        data = ujson.dumps({
+            "ProfilePhotoURL": pfp,
+            "UsernameColor": colour,
+            "Username": username,
+            "AdminTitle": admintitle,
+            "Text": reply.raw_text,
```
Comment (penn5) [inline quotes.py]: That should be .message

File: quotes.py [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+            "Markdown": get_markdown(reply),
+            "Template": args[0].lower(),
+            "APIKey": self.config["API_TOKEN"]
+        })
+        requested = requests.post(f'{self.config["API_URL"]}/api/v2/quote', data=data)
+
+        requested = requested.json()
```
Comment (penn5) [inline quotes.py]: This also needs to be async

File: quotes.py [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+
+        if requested['status'] == 500:
```
Comment (penn5) [inline quotes.py]: No. If an error happens, it should be an error. Use .raise_for_status(), assuming the server sends the http status the same as the JSON status

File: quotes.py [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+            return await utils.answer(message, 'Server error, please report to the developer')
+        elif requested['status'] == 401:
+            if requested['message'] == 'ERROR_TOKEN_INVALID':
+                return await utils.answer(message, '<code>Wrong API Token.</code>')
```
Comment (penn5) [inline quotes.py]: Use bold

File: quotes.py [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+        elif requested['status'] == 403:
+            if requested['message'] == 'ERROR_UNAUTHORIZED':
+                return await utils.answer(message, '<code>You don\'t have enough permissions.</code>')
+        elif requested['status'] == 404:
+            if requested['message'] == 'ERROR_TEMPLATE_NOT_FOUND':
+                newreq = requests.post(f'{self.config["API_URL"]}/api/v1/getalltemplates', data={
```
Comment (penn5) [inline quotes.py]: Format strings are not appropriate

Comment (penn5) [inline quotes.py]: Format strings are not appropriate

File: quotes.py [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+                    "token": self.config["API_TOKEN"]
+                }).json()
```
Comment (penn5) [inline quotes.py]: Use continued indentation

File: quotes.py [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+
+                if newreq['status'] == 'NOT_ENOUGH_PERMISSIONS':
+                    return await utils.answer(message, '<code>Wrong template. You can use only default</code>')
```
Comment (penn5) [inline quotes.py]: Hmm.

File: quotes.py [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+                else:
+                    templates = ', '.join(newreq['message'])
```
Comment (penn5) [inline quotes.py]: Make that translateable with formatting, and use ""

File: quotes.py [hunk 28]
```diff
@@ -0,0 +1,258 @@
...
+                    return await utils.answer(message, f'<code>Wrong template. Available: {templates}</code>')
+
+        file = None
+        try:
+            req = requests.get(f'{self.config["API_URL"]}/cdn/' + requested['message'])
```
Comment (penn5) [inline quotes.py]: No f-strings here

File: quotes.py [hunk 29]
```diff
@@ -0,0 +1,258 @@
...
+            file = BytesIO(req.content)
```
Comment (penn5) [inline quotes.py]: WTH‽

File: quotes.py (hunk 28 above)
Comment (penn5) [reply quotes.py]: Asyncio

File: quotes.py (hunk 29 above)
Comment (penn5) [reply quotes.py]: Also asyncio

File: quotes.py [hunk 30]
```diff
@@ -0,0 +1,258 @@
...
+            file.seek(0)
+        except HTTPError:
+            return await utils.answer(message, 'Trolling occurred')
```
Comment (penn5) [inline quotes.py]: Wtf

File: quotes.py [hunk 31]
```diff
@@ -0,0 +1,258 @@
...
+                    return await utils.answer(message, '<code>Wrong template. You can use only default</code>')
+                else:
+                    templates = ', '.join(newreq['message'])
+                    return await utils.answer(message, f'<code>Wrong template. Available: {templates}</code>')
+
+        file = None
+        try:
+            req = requests.get(f'{self.config["API_URL"]}/cdn/' + requested['message'])
+            file = BytesIO(req.content)
+            file.seek(0)
```
Comment (penn5) [inline quotes.py]: Not relevant

File: quotes.py [hunk 32]
```diff
@@ -0,0 +1,258 @@
...
+        except HTTPError:
+            return await utils.answer(message, 'Trolling occurred')
+
+        await msg.delete()
```
Comment (penn5) [inline quotes.py]: No. Use utils.answer.

File: quotes.py [hunk 33]
```diff
@@ -0,0 +1,258 @@
...
+        if len(args) == 1 or args[1].lower() == 'sticker':
+            sticker = BytesIO()
+            Image.open(file).save(sticker, 'webp')
+            sticker.name = "sticker.webp"
+            sticker.seek(0)
+            await self.client.send_file(message.chat_id, sticker, reply_to=reply.id, file_name="sticker.webp")
```
Comment (penn5) [inline quotes.py]: Utils.answer

File: quotes.py [hunk 34]
```diff
@@ -0,0 +1,258 @@
...
+            req = requests.get(f'{self.config["API_URL"]}/cdn/' + requested['message'])
+            file = BytesIO(req.content)
+            file.seek(0)
+        except HTTPError:
+            return await utils.answer(message, 'Trolling occurred')
+
+        await msg.delete()
+        if len(args) == 1 or args[1].lower() == 'sticker':
+            sticker = BytesIO()
+            Image.open(file).save(sticker, 'webp')
```
Comment (penn5) [inline quotes.py]: Asyncio for resource intensive operations.

File: quotes.py [hunk 35]
```diff
@@ -0,0 +1,258 @@
...
+            sticker.name = "sticker.webp"
+            sticker.seek(0)
+            await self.client.send_file(message.chat_id, sticker, reply_to=reply.id, file_name="sticker.webp")
+            sticker.close()
```
Comment (penn5) [inline quotes.py]: Use a context manager

File: quotes.py [hunk 36]
```diff
@@ -0,0 +1,258 @@
...
+        else:
+            test = args[1].lower()
```
Comment (penn5) [inline quotes.py]: Inline that.

File: quotes.py [hunk 37]
```diff
@@ -0,0 +1,258 @@
...
+            file.name = "file.png"
```
Comment (penn5) [inline quotes.py]: Maybe make that modifiable? Not sure

File: quotes.py [hunk 38]
```diff
@@ -0,0 +1,258 @@
...
+                             MessageEntityMentionName, MessageEntityHashtag,
+                             MessageEntityCashtag, MessageEntityBotCommand,
+                             MessageEntityUrl]:
+            md_item["Type"] = "link"
+        elif entity_type is MessageEntityCode:
+            md_item["Type"] = "code"
+        elif entity_type is MessageEntityStrike:
+            md_item["Type"] = "stroke"
+        elif entity_type is MessageEntityUnderline:
+            md_item["Type"] = "underline"
```
Comment (penn5) [inline quotes.py:230]: Log a warning if it's not matched

File: quotes.py [hunk 39]
```diff
@@ -0,0 +1,258 @@
...
+
+def get_markdown(reply):
+    if not reply.entities:
+        return []
+
+    markdown = []
+    for entity in reply.entities:
+        entity_type = type(entity)
+        md_item = {
+            "Type": "",
```
Comment (penn5) [inline quotes.py]: Better to initialise to None, as it's a singleton

File: quotes.py [hunk 40]
```diff
@@ -0,0 +1,258 @@
...
+        else:
+            msg = await message.edit('<code>Processing...</code>')
+
+        username = ''
+        no_pfp = True
+        colour = '#b48bf2'
+        admintitle = ''
+        username = ''
+        pfp = ''
+        user_id = 000000000
```
Comment (penn5) [inline quotes.py]: Wtf is this crap?

File: quotes.py (hunk 4 above)
Comment (penn5) [reply quotes.py]: See afk or nopm for examples

File: quotes.py [hunk 41]
```diff
@@ -0,0 +1,258 @@
...
+            msg = await message.edit('<code>Processing... Note: only message will get processed.</code>')
+        else:
+            msg = await message.edit('<code>Processing...</code>')
+
+        username = ''
+        no_pfp = True
+        colour = '#b48bf2'
+        admintitle = ''
+        username = ''
+        pfp = ''
```
Comment (penn5) [inline quotes.py]: Ummmmmmm. Afaik you can't have a string for your photo. None is more appropriate.

File: quotes.py [hunk 42]
```diff
@@ -0,0 +1,258 @@
...
+        if not args:
+            return await utils.answer(message, '<code>You didn\'t specify template</code>')
+
+        if reply.media:
+            msg = await message.edit('<code>Processing... Note: only message will get processed.</code>')
+        else:
+            msg = await message.edit('<code>Processing...</code>')
+
+        username = ''
+        no_pfp = True
```
Comment (penn5) [inline quotes.py]: Why does that even exist? Just check truthiness of the pfp

File: quotes.py [hunk 43]
```diff
@@ -0,0 +1,258 @@
...
+
+        if not args:
+            return await utils.answer(message, '<code>You didn\'t specify template</code>')
+
+        if reply.media:
+            msg = await message.edit('<code>Processing... Note: only message will get processed.</code>')
+        else:
+            msg = await message.edit('<code>Processing...</code>')
+
+        username = ''
```
Comment (penn5) [inline quotes.py]: Again use None. That way you get errors thrown if there's bad logic.

File: quotes.py (hunk 5 above)
Comment (penn5) [reply quotes.py]: As a result L84-130 will need refactoring.

File: quotes.py [hunk 44]
```diff
@@ -0,0 +1,258 @@
...
+            msg = await message.edit("<code>Processing...</code>")
+
+        username = ""
+        no_pfp = True
+        colour = "#b48bf2"
+        admintitle = ""
+        username = ""
+        pfp = ""
+        user_id = 000000000
+        fwd_status = "no_fwd"
```
Comment (penn5) [inline quotes.py]: Inline this variable. That means rewriting L84-130

File: quotes.py [hunk 45]
```diff
@@ -0,0 +1,258 @@
...
+        elif fwd_status == "hidden":
+            username = reply.fwd_from.from_name
+        elif fwd_status == "channel_private":
+            username = "Hidden."
+        else:
+            username = user.first_name
+            if user.last_name:
+                username += f" {user.last_name}"
+
+        if fwd_status.startswith("channel"):
```
Comment (penn5) [inline quotes.py]: That smells.

File: quotes.py [hunk 46]
```diff
@@ -0,0 +1,258 @@
...
+            "ProfilePhotoURL": pfp,
+            "UsernameColor": colour,
+            "Username": username,
+            "AdminTitle": admintitle,
+            "Text": reply.raw_text,
+            "Markdown": get_markdown(reply),
+            "Template": args[0].lower(),
+            "APIKey": self.config["API_TOKEN"]
+        })
+        requested = requests.post(self.config["API_URL"] + "/api/v2/quote", data=data)
```
Comment (penn5) [inline quotes.py]: missing async

File: quotes.py [hunk 47]
```diff
@@ -0,0 +1,258 @@
...
+
+        requested = requested.json()
+
+        if requested["status"] == 500:
+            return await utils.answer(message, "Server error, please report to the developer")
```
Comment (penn5) [inline quotes.py]: This should all be an exception. Use .raise_for_status()

File: quotes.py [hunk 48]
```diff
@@ -0,0 +1,258 @@
...
+        elif requested["status"] == 401:
+            if requested["message"] == "ERROR_TOKEN_INVALID":
+                return await utils.answer(message, "<code>Wrong API Token.</code>")
+        elif requested["status"] == 403:
+            if requested["message"] == "ERROR_UNAUTHORIZED":
+                return await utils.answer(message, "<code>You don\"t have enough permissions.</code>")
+        elif requested["status"] == 404:
+            if requested["message"] == "ERROR_TEMPLATE_NOT_FOUND":
+                newreq = requests.post(self.config["API_URL"] + "/api/v1/getalltemplates", data={
```
Comment (penn5) [inline quotes.py]: Async!

File: quotes.py [hunk 49]
```diff
@@ -0,0 +1,258 @@
...
+        })
+        requested = requests.post(self.config["API_URL"] + "/api/v2/quote", data=data)
+
+        requested = requested.json()
+
+        if requested["status"] == 500:
+            return await utils.answer(message, "Server error, please report to the developer")
+        elif requested["status"] == 401:
+            if requested["message"] == "ERROR_TOKEN_INVALID":
+                return await utils.answer(message, "<code>Wrong API Token.</code>")
```
Comment (penn5) [inline quotes.py]: Use bold not code

File: quotes.py [hunk 50]
```diff
@@ -0,0 +1,258 @@
...
+            if requested["message"] == "ERROR_UNAUTHORIZED":
+                return await utils.answer(message, "<code>You don\"t have enough permissions.</code>")
+        elif requested["status"] == 404:
+            if requested["message"] == "ERROR_TEMPLATE_NOT_FOUND":
+                newreq = requests.post(self.config["API_URL"] + "/api/v1/getalltemplates", data={
+                    "token": self.config["API_TOKEN"]
+                }).json()
+
+                if newreq["status"] == "NOT_ENOUGH_PERMISSIONS":
+                    return await utils.answer(message, "<code>Wrong template. You can use only default</code>")
```
Comment (penn5) [inline quotes.py]: Eh?

File: quotes.py [hunk 51]
```diff
@@ -0,0 +1,258 @@
...
+                else:
+                    templates = ", ".join(newreq["message"])
```
Comment (penn5) [inline quotes.py]: Make that translateable

File: quotes.py [hunk 52]
```diff
@@ -0,0 +1,258 @@
...
+                    return await utils.answer(message, f"<code>Wrong template. Available: {templates}</code>")
+
+        file = None
+        try:
+            req = requests.get(self.config["API_URL"] + "/cdn/" + requested["message"])
+            file = BytesIO(req.content)
+            file.seek(0)
+        except HTTPError:
+            return await utils.answer(message, "Trolling occurred")
```
Comment (penn5) [inline quotes.py]: ?????

File: quotes.py [hunk 53]
```diff
@@ -0,0 +1,258 @@
...
+
+        await msg.delete()
+        if len(args) == 1 or args[1].lower() == "sticker":
+            sticker = BytesIO()
+            Image.open(file).save(sticker, "webp")
```
Comment (penn5) [inline quotes.py]: Use async on cpu-intensive stuff

File: quotes.py [hunk 54]
```diff
@@ -0,0 +1,258 @@
...
+            sticker.seek(0)
+            await self.client.send_file(message.chat_id, sticker, reply_to=reply.id, file_name="sticker.webp")
+            sticker.close()
+        else:
+            test = args[1].lower()
+            file.name = "file.png"
+            if test == "file":
+                await self.client.send_file(message.chat_id, file, reply_to=reply.id, file_name="file.png")
+            elif test == "force_file":
+                await self.client.send_file(message.chat_id, file, reply_to=reply.id,
```
Comment (penn5) [inline quotes.py]: Utils.answer

File: quotes.py [hunk 55]
```diff
@@ -0,0 +1,258 @@
...
+            Image.open(file).save(sticker, "webp")
+            sticker.name = "sticker.webp"
+            sticker.seek(0)
+            await self.client.send_file(message.chat_id, sticker, reply_to=reply.id, file_name="sticker.webp")
+            sticker.close()
+        else:
+            test = args[1].lower()
+            file.name = "file.png"
+            if test == "file":
+                await self.client.send_file(message.chat_id, file, reply_to=reply.id, file_name="file.png")
```
Comment (penn5) [inline quotes.py]: Utils.answer

File: quotes.py [hunk 56]
```diff
@@ -0,0 +1,258 @@
...
+            elif test == "force_file":
+                await self.client.send_file(message.chat_id, file, reply_to=reply.id,
+                                            force_document=True, file_name="file.png")
+        file.close()
```
Comment (penn5) [inline quotes.py]: Context manager

File: quotes.py (hunk 42 above)
Comment (pqhaz) [reply quotes.py]: ye its not required at all in new api, imma remove that

File: quotes.py (hunk 22 above)
Comment (pqhaz) [reply quotes.py]: no it doesn't, it sends 200 in all conditions

Comment (penn5) [reply quotes.py]: Well that's just stupid

File: quotes.py [hunk 57]
```diff
@@ -0,0 +1,234 @@
...
+from io import BytesIO
+from urllib.error import HTTPError
+
+logger = logging.getLogger(__name__)
+
+
+def register(cb):
+    cb(QuotesMod())
+
+
```
Comment (penn5) [inline quotes.py:35]: @loader.tds

File: quotes.py [hunk 58]
```diff
@@ -0,0 +1,233 @@
...
+class QuotesMod(loader.Module):
+    """Quote a message."""
+    strings = {
+        "name": "Quotes",
+        "API_TOKEN": "API Key/Token for Quotes.",
```
Comment (penn5) [inline quotes.py]: make all the keys for this lowercase

Comment (penn5) [reply quotes.py]: and the config ones should be called `<thing>_cfg_doc`

File: quotes.py [hunk 59]
```diff
@@ -0,0 +1,233 @@
...
+        self.name = self.strings["name"]
+
+    async def client_ready(self, client, db):
+        self.client = client
+
+    async def quotecmd(self, message):  # noqa: C901
+        """Quote a message.
+        Usage: .quote [template]
+        If template is missing, possible templates are fetched."""
+
```
Comment (penn5) [inline quotes.py]: redundant newline

File: quotes.py [hunk 60]
```diff
@@ -0,0 +1,233 @@
...
+        args = utils.get_args(message)
+        reply = await message.get_reply_message()
+
+        if not reply:
+            return await utils.answer(message, self.strings["no_reply"])
+
+        if not args:
+            return await utils.answer(message, self.strings["no_template"])
+
+        ProfilePhotoURL = UsernameColor = Username = AdminTitle = UserID = Channel = None
```
Comment (penn5) [inline quotes.py]: PYTHONIC NAMING

File: quotes.py [hunk 61]
```diff
@@ -0,0 +1,233 @@
...
+
+        if not reply.fwd_from:
+            user = await self.client(telethon.tl.functions.channels.GetParticipantRequest(message.chat_id,
+                                                                                          reply.from_id))
+            user_full = await self.client(telethon.tl.functions.users.GetFullUserRequest(reply.from_id))
+            if not await self.client.get_profile_photos(reply.from_id):
+                ProfilePhotoURL = ""
+
+            AdminTitle = getattr(user.participant, "participant", "")
+            if AdminTitle is None:
```
Comment (penn5) [inline quotes.py]: This can be replaced with `admin_title = getattr(user.participant, "participant", None) or ""`

File: quotes.py [hunk 62]
```diff
@@ -0,0 +1,233 @@
...
+                        ProfilePhotoURL = ""
+
+                    AdminTitle = "Channel"
+                    Username = telethon.utils.get_display_name(resolved)
+                    UserID = cid
+                    Channel = resolved
+                except telethon.errors.rpcerrorlist.ChannelPrivateError:
+                    # we don't know anything about this channel.
+                    ProfilePhotoURL = ""
+                    Username = "Hidden."
```
Comment (penn5) [inline quotes.py]: TRANSLATIONS

File: quotes.py [hunk 63]
```diff
@@ -0,0 +1,233 @@
...
+                AdminTitle = "Channel"  # 100%
```
Comment (penn5) [inline quotes.py]: TRANSLATIONS

File: quotes.py [hunk 64]
```diff
@@ -0,0 +1,233 @@
...
+                Username = reply.fwd_from.from_name
+            elif reply.fwd_from.channel_id:
+                try:
+                    cid = reply.fwd_from.channel_id
+                    resolved = (await self.client(telethon.tl.functions.channels.GetFullChannelRequest(cid))).chats[0]
+
+                    if isinstance(resolved.photo, telethon.tl.types.ChatPhotoEmpty):  # get_profile_photos won't work
+                        ProfilePhotoURL = ""
+
+                    AdminTitle = "Channel"
```
Comment (penn5) [inline quotes.py]: translations

File: quotes.py [hunk 65]
```diff
@@ -0,0 +1,233 @@
...
+                    resolved = (await self.client(telethon.tl.functions.users.GetFullUserRequest(UserID))).user
+
+                    if not await self.client.get_profile_photos(UserID):
+                        ProfilePhotoURL = ""
+
+                    Username = telethon.utils.get_display_name(resolved)
+                    AdminTitle = ""
+                except ValueError:
+                    ProfilePhotoURL = ""
+                    Username = "A user."
```
Comment (penn5) [inline quotes.py]: translations

File: quotes.py [hunk 66]
```diff
@@ -0,0 +1,233 @@
...
+                    AdminTitle = "User"
```
Comment (penn5) [inline quotes.py]: trnsl

File: quotes.py [hunk 67]
```diff
@@ -0,0 +1,233 @@
...
+
+        if ProfilePhotoURL != "" or Channel:
+            with BytesIO() as pfp:
+                if Channel:
+                    await self.client.download_profile_photo(Channel, pfp)
```
Comment (penn5) [inline quotes.py]: If you're using download_profile_photo, it's more efficient do `download_profile_photo(..., bytes)` and it returns a `bytes` object

File: quotes.py [hunk 68]
```diff
@@ -0,0 +1,233 @@
...
+                else:
+                    await self.client.download_profile_photo(UserID, pfp)
+                pfp.seek(0)
+                ProfilePhotoURL = "data:image/png;base64, " + base64.b64encode(pfp.read()).decode()
+                pfp.close()
```
Comment (penn5) [inline quotes.py]: Redundant

File: quotes.py [hunk 69]
```diff
@@ -0,0 +1,233 @@
...
+            "Text": reply.message,
+            "Markdown": get_markdown(reply),
+            "Template": args[0],
+            "APIKey": self.config["API_TOKEN"]
+        })
+
+        resp = await utils.run_sync(requests.post, self.config["API_URL"] + "/api/v2/quote", data=request)
+        resp = await utils.run_sync(resp.json)
+
+        if resp["status"] == 500:  # API RETURNS 200 STATUS CODE ALL TIME
```
Comment (penn5) [inline quotes.py]: Does the comment have to be quite so shouty?

File: quotes.py [hunk 70]
```diff
@@ -0,0 +1,233 @@
...
+            "UsernameColor": UsernameColor,
+            "Username": Username,
+            "AdminTitle": AdminTitle,
+            "Text": reply.message,
+            "Markdown": get_markdown(reply),
+            "Template": args[0],
+            "APIKey": self.config["API_TOKEN"]
+        })
+
+        resp = await utils.run_sync(requests.post, self.config["API_URL"] + "/api/v2/quote", data=request)
```
Comment (penn5) [inline quotes.py]: put between these 2 lines, `resp.raise_for_status()` (doesn't need async magic)

File: quotes.py [hunk 71]
```diff
@@ -0,0 +1,233 @@
...
+        resp = await utils.run_sync(resp.json)
+
+        if resp["status"] == 500:  # API RETURNS 200 STATUS CODE ALL TIME
+            return await utils.answer(message, self.strings["SERVER_ERROR"])
+        elif resp["status"] == 401:
+            if resp["message"] == "ERROR_TOKEN_INVALID":
```
Comment (penn5) [inline quotes.py:158]: This needs an `else`

File: quotes.py [hunk 72]
```diff
@@ -0,0 +1,233 @@
...
+                return await utils.answer(message, self.strings["INVALID_TOKEN"])
+        elif resp["status"] == 403:
+            if resp["message"] == "ERROR_UNAUTHORIZED":
```
Comment (penn5) [inline quotes.py:163]: `else`

File: quotes.py [hunk 73]
```diff
@@ -0,0 +1,233 @@
...
+                return await utils.answer(message, self.strings["UNAUTHORIZED"])
+        elif resp["status"] == 404:
+            if resp["message"] == "ERROR_TEMPLATE_NOT_FOUND":
```
Comment (penn5) [inline quotes.py:168]: `else`

File: quotes.py [hunk 74]
```diff
@@ -0,0 +1,233 @@
...
+                newreq = await utils.run_sync(requests.post, self.config["API"] + "/api/v1/getalltemplates", data={
+                    "token": self.config["API_TOKEN"]
+                })
+                newreq = await utils.run_sync(newreq.json)
+
+                if newreq["status"] == "NOT_ENOUGH_PERMISSIONS":
```
Comment (penn5) [inline quotes.py:174]: Isn't status numerical?

File: quotes.py [hunk 75]
```diff
@@ -0,0 +1,233 @@
...
+                    return await utils.answer(message, self.strings["NOT_ENOUGH_PERMISSIONS"])
+                else:
+                    templates = self.strings["DELIMITER"].join(newreq["message"])
```
Comment (penn5) [inline quotes.py]: not like that.
`templates = "".join(self.strings["template_format"].format(template) for template in newreq["message"])`

File: quotes.py (hunk 29 above)
Comment (penn5) [reply quotes.py]: You ought to use aiohttp here tbh. Else you will hit issues.

File: quotes.py [hunk 76]
```diff
@@ -0,0 +1,233 @@
...
+        except HTTPError:
+            return await utils.answer(message, self.strings["CANNOT_REACH_CDN"])
+
+        img = await utils.run_sync(Image.open, file)
+        with BytesIO() as sticker:
+            await utils.run_sync(img.save, sticker, "webp")
+            sticker.name = "sticker.webp"
+            sticker.seek(0)
+            await utils.answer(message, sticker)
+            sticker.close()
```
Comment (penn5) [inline quotes.py]: redundant close

File: quotes.py (hunk 61 above)
Comment (penn5) [reply quotes.py]: I feel like you forgot to get the .rank?

File: quotes.py (hunk 69 above)
Comment (penn5) [reply quotes.py]: This if needs an `elif resp["status"] != 200`

File: quotes.py [hunk 77]
```diff
@@ -0,0 +1,233 @@
...
+                newreq = await utils.run_sync(newreq.json)
+
+                if newreq["status"] == "NOT_ENOUGH_PERMISSIONS":
+                    return await utils.answer(message, self.strings["NOT_ENOUGH_PERMISSIONS"])
+                else:
+                    templates = self.strings["DELIMITER"].join(newreq["message"])
+                    return await utils.answer(message, self.strings["TEMPLATES"].format(templates))
+
+        try:
+            req = await utils.run_sync(requests.get, self.config["API_URL"] + "/cdn/" + resp["message"])
```
Comment (penn5) [inline quotes.py]: use .raise_for_status()

File: quotes.py [hunk 78]
```diff
@@ -0,0 +1,233 @@
...
+            file = BytesIO(req.content)
+            file.seek(0)
+        except HTTPError:
```
Comment (penn5) [inline quotes.py]: afaik that's never raised.

File: quotes.py (hunk 74 above)
Comment (pqhaz) [reply quotes.py]: no, here it isn't

File: quotes.py (hunk 69 above)
Comment (pqhaz) [reply quotes.py]: nooooooooo then it won't handle other errors

File: quotes.py (hunk 67 above)
Comment (pqhaz) [reply quotes.py]: do it yourself if you want to, I'm not going to

File: quotes.py (hunk 72 above)
Comment (pqhaz) [reply quotes.py]: what?

File: quotes.py (hunk 71 above)
Comment (pqhaz) [reply quotes.py]: what?

**Comment by pqhaz** [comment]
brb, not going to resolve more

**Comment by pqhaz** [comment]
brb, not going to resolve more

**Pull request merged**